### ISSUE TITLE
Add --staged flag to only check changes in staged code

### ISSUE THREAD

**Issue by abemedia** [golangci/golangci-lint#4328]
### Your feature request related to a problem? Please describe.

I use golangci-lint in pre-commit hooks and without adding a custom script to first stash unstaged changes this results in the pre-commit hook failing if there are issues in unstaged code.

### Describe the solution you'd like.

A new flag `--staged` which ignores all issues that are not in staged changes.

### Describe alternatives you've considered.

A custom pre-commit script which first stashes the changes e.g.
```sh
git stash --keep-index --include-untracked --quiet
exitCode=0
golangci-lint run || exitCode=$?
git stash pop --quiet
exit $exitCode
```
The problem with this solution is that it can result in the loss of what is currently staged and what isn't.

**Comment by ttlins** [comment]
hey @abemedia , are there any current proposals on how to tackle that? I'm looking for this exact same solution, would be glad to work on PR later if you haven't started yet 👍

**Comment by abemedia** [comment]
Well there's already logic to just use a patch file. One could implement this by generating a patch file from the currently staged changes.

**Comment by ttlins** [comment]
ah, I wasn't aware of the `new-from-patch` flag. That's easier than I thought then 🙌 
yeah, I'd suggest using the `git diff ...` approach also. I'll see if I can work on that later today/tomorrow then 👍 
Thanks!

**Comment by ldez** [comment]
It was not really a suggestion but more a technical thinking.

I'm not sure of the value of an option only for a niche that can be done easily with the command line.
This is why I flagged this issue with `no-decision`.

```bash
git diff --cached > stage.patch; golangci-lint run --new-from-patch=stage.patch; rm stage.patch
``

**Comment by ttlins** [comment]
> I'm not sure of the value of an option, only for a niche, that can be done easily with the command line.

Yeah, to be honest I've had the same thought. After reading your comments, I re-wrote my pre commit hook to do pretty much the same as you've suggested above. 

I'm just sad since this was an easy first contribution 😅 

> Also, note that running golangci-lint only on a few files can lead to unexpected behavior with linters (the majority) that require scanning all files.

I wasn't aware of this though - thanks for raising it 👍

**Comment by NBS-LLC** [comment]
I think a `--staged` flag would still be useful. For example, in a monorepo where I have a number of go services that are sub-projects. I'd like to use golangci-lint in a husky pre-commit hook that runs against each sub-project only if there are staged changes.

Ideally I should be able to use this in conjunction with `--whole-files`.
